### ISSUE TITLE
Unable to connect with UnixODBC and Cloudera Impala Driver in pyodbc 5.0.1

### ISSUE THREAD

**Issue by Deividhp13** [mkleehammer/pyodbc#1313]
Hello everyone, first of all thanks for the hard work, I've been using this library for so long and it's awesome.

Since the last update to pyodbc version 5, the connection with UnixODBC and the Cloudera Impala driver has stopped working.  I am making the connection directly with DSN, configuring /etc/odbc.ini. 

**I've tried with latest v4, and it's working as expected**, but with the new release I'm getting the following error:

`InterfaceError: ('IM002', '[IM002] [unixODBC][Driver Manager]Data source name not found, and no default driver specified (0) (SQLDriverConnect)')`

![image](https://github.com/mkleehammer/pyodbc/assets/23059822/65fbf41c-2c8b-4b5d-98f5-b50c08b09aa5)

This is my odbc.ini file:

```
[ODBC Data Sources]
Hadoop LDAP=Cloudera ODBC Driver for Impala
[Hadoop LDAP]
Driver=/opt/cloudera/impalaodbc/lib/64/libclouderaimpalaodbc64.so
Host=<myHost>
Port=25008
AuthMech=User Name and Password
UID=<myUID>
PWD=<myPWD>
```

And this is how I'm configuring the connection:
```
import pyodbc
conn = pyodbc.connect('DSN=Hadoop LDAP')
cursor = conn.cursor()
```

I've tried also to remove white spaces on DSN name, but the error persists.

Thanks!

**Comment by gordthompson** [comment]
Can you reproduce the issue from the command line? For example

```
(py312) gord@xubu-22-04:~$ python
Python 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 17:29:18) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyodbc
>>> pyodbc.version
'5.0.1'
>>> conn = pyodbc.connect("DSN=Hadoop LDAP")
>>>

**Comment by Deividhp13** [comment]
> Can you reproduce the issue from the command line? For example
> 
> ```
> (venv) gord@xubu-22-04:~$ python
> Python 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 17:29:18) [GCC 11.2.0] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import pyodbc
> >>> pyodbc.version
> '5.0.1'
> >>> conn = pyodbc.connect("DSN=Hadoop LDAP")
> >>> 
> ```

Yes, from the command line I get exactly the same error. I'm working with Python 3.10.11.

```
(venv) [adm-herreda1@idcsrv404 gda_dbio]$ python
Python 3.10.11 (main, Jun 19 2023, 11:38:26) [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyodbc
>>> pyodbc.version
'5.0.1'
>>> conn = pyodbc.connect("DSN=Hadoop Catalyst LDAP")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
pyodbc.InterfaceError: ('IM002', '[IM002] [unixODBC][Driver Manager]Data source name not found, and no default driver specified (0) (SQLDriverConnect)')
```

I tried with both versions, and on 4.0.39 it's working on console.

```
Python 3.10.11 (main, Jun 19 2023, 11:38:26) [GCC 4.8.5 20150623 (Red Hat 4.8.5-44)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyodbc
>>> pyodbc.version
'4.0.39'
>>> conn = pyodbc.connect("DSN=Hadoop Catalyst LDAP", autocommit=True)
>>> conn
<pyodbc.Connection object at 0x7fa404b73210>
```

On windows, with version 5.0.1 it's working fine, so the problem might be related with linux somehow. 

```
Python 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyodbc                                                     
>>> pyodbc.version
'5.0.1'
>>> conn = pyodbc.connect("DSN=Hadoop Catalyst LDAP", autocommit = True)                                                                    
>>> conn
<pyodbc.Connection object at 0x000001E26F1D3910>
```

**Comment by gordthompson** [comment]
Okay, so let's try and determine what pyodbc 5.0.1 is doing differently from 4.0.39.

Create a little test script

```python
# pyodbc_test.py

import pyodbc

print(pyodbc.version)
conn = pyodbc.connect("DSN=DSN=Hadoop Catalyst LDAP", autocommit=True)
```

verify that pyodbc 4.0.39 is installed, then run this from the command prompt

```
strace -o pyodbc_4.txt python pyodbc_test.py
```

Upgrade pyodbc to 5.0.1 and then run

```
strace -o pyodbc_5.txt python pyodbc_test.py
```

Attach (drag and drop) those .txt files into a reply here and we can see if anything looks different.

**Comment by v-chojas** [comment]
Could you try an ODBC trace instead?

**Comment by v-chojas** [comment]
From the straces I see your /etc/odbcinst.ini appears to be empty, which doesn't make sense - that's where the driver should be listed.

**Comment by Spitihnev** [comment]
My /etc/odbcinst.ini is indeed empty but the 4.x version doesn't mind. How do I run ODBC trace? Tried setting trace options in odbc.ini under  [ODBC] but the log file is nowhere to be found. The tracing under my DSN_name was already set as is.

my odbc.ini file looks like this:
`[ODBC]
# Specify any global ODBC configuration here such as ODBC tracing.
Trace=Yes
TraceFile=/tmp/odbc.log

[ODBC Data Sources]
DSN_name=Impala

[DSN_name]
Description=Cloudera ODBC Driver for Impala (64-bit) DSN

# Driver: The location where the ODBC driver is installed to.
Driver=/opt/cloudera/impalaodbc/lib/64/libclouderaimpalaodbc64.so

HOST=<my_host>
PORT=21050
Schema=my_schema
ErrorMessagesPath=/home/root/ErrorMessage/
Trace=Yes
TraceFile=/tmp/odbc.log
LogLevel=4
LogPath=/tmp

EnableSimulatedTransactions=1
AuthMech=1

# Kerberos related settings.
KrbFQDN=<kerberos_host>

CAIssuedCertNamesMismatch=1
TrustedCerts=/opt/cloudera/impalaodbc/lib/64/cacerts.pem`

**Comment by gordthompson** [comment]
The

```
[ODBC]
Trace=Yes
TraceFile=/tmp/odbc.log
```

block goes in odbcinst.ini. Putting it in odbc.ini has no effect.

**Comment by Spitihnev** [comment]
Thanks here is the logfile, 1st 5.x version connect and then 4.x

```
[ODBC][3460][1712343225.497105][__handles.c][460]
		Exit:[SQL_SUCCESS]
			Environment = 0x705990
[ODBC][3460][1712343225.497153][SQLSetEnvAttr.c][189]
		Entry:
			Environment = 0x705990
			Attribute = SQL_ATTR_ODBC_VERSION
			Value = 0x3
			StrLen = 4
[ODBC][3460][1712343225.497165][SQLSetEnvAttr.c][363]
		Exit:[SQL_SUCCESS]
[ODBC][3460][1712343225.497178][SQLAllocHandle.c][375]
		Entry:
			Handle Type = 2
			Input Handle = 0x705990
[ODBC][3460][1712343225.497188][SQLAllocHandle.c][493]
		Exit:[SQL_SUCCESS]
			Output Handle = 0x6fc3d0
[ODBC][3460][1712343225.497479][SQLDriverConnectW.c][290]
		Entry:
			Connection = 0x6fc3d0
			Window Hdl = (nil)
			Str In = [DSN=my_dsn][length = 15 (SQL_NTS)]
			Str Out = (nil)
			Str Out Max = 0
			Str Out Ptr = (nil)
			Completion = 0
[ODBC][3460][1712343225.498389][SQLDriverConnectW.c][535]Error: IM002
[ODBC][3460][1712343225.498416][SQLGetDiagRecW.c][508]
		Entry:
			Connection = 0x6fc3d0
			Rec Number = 1
			SQLState = 0x7ffd262cdba4
			Native = 0x7ffd262cdb7c
			Message Text = 0x711550
			Buffer Length = 1023
			Text Len Ptr = 0x7ffd262cdb7a
[ODBC][3460][1712343225.498433][SQLGetDiagRecW.c][550]
		Exit:[SQL_SUCCESS]
			SQLState = IM002
			Native = 0x7ffd262cdb7c -> 0
			Message Text = [[unixODBC][Driver Manager]Data source name not found, and no default driver specified]
[ODBC][3460][1712343225.498471][SQLFreeHandle.c][284]
		Entry:
			Handle Type = 2
			Input Handle = 0x6fc3d0
[ODBC][3460][1712343225.498485][SQLFreeHandle.c][333]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.015005][__handles.c][460]
		Exit:[SQL_SUCCESS]
			Environment = 0x23fa860
[ODBC][3497][1712343246.015037][SQLSetEnvAttr.c][189]
		Entry:
			Environment = 0x23fa860
			Attribute = SQL_ATTR_ODBC_VERSION
			Value = 0x3
			StrLen = 4
[ODBC][3497][1712343246.015048][SQLSetEnvAttr.c][363]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.015061][SQLAllocHandle.c][375]
		Entry:
			Handle Type = 2
			Input Handle = 0x23fa860
[ODBC][3497][1712343246.015070][SQLAllocHandle.c][493]
		Exit:[SQL_SUCCESS]
			Output Handle = 0x23f12a0
[ODBC][3497][1712343246.015340][SQLDriverConnectW.c][290]
		Entry:
			Connection = 0x23f12a0
			Window Hdl = (nil)
			Str In = [DSN=my_dsn][length = 15 (SQL_NTS)]
			Str Out = (nil)
			Str Out Max = 0
			Str Out Ptr = (nil)
			Completion = 0
[ODBC][3497][1712343246.015845][SQLDriverConnectW.c][535]Error: IM002
[ODBC][3497][1712343246.015879][SQLDriverConnect.c][726]
		Entry:
			Connection = 0x23f12a0
			Window Hdl = (nil)
			Str In = [DSN=my_dsn][length = 15 (SQL_NTS)]
			Str Out = 0x7ffd61346c10
			Str Out Max = 2048
			Str Out Ptr = (nil)
			Completion = 0
		UNICODE Using encoding ASCII 'ISO8859-1' and UNICODE 'UCS-2LE'

[ODBC][3497][1712343246.059454][SQLDriverConnect.c][1582]
		Exit:[SQL_SUCCESS]
			Connection Out [[DSN=my_dsn;][length = 16 (SQL_NTS)]]
[ODBC][3497][1712343246.059584][SQLGetInfo.c][554]
		Entry:
			Connection = 0x23f12a0
			Info Type = SQL_DRIVER_ODBC_VER (77)
			Info Value = 0x7ffd61347c10
			Buffer Length = 20
			StrLen = 0x7ffd61347bf0
[ODBC][3497][1712343246.059617][SQLGetInfo.c][617]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.059630][SQLGetInfo.c][554]
		Entry:
			Connection = 0x23f12a0
			Info Type = SQL_DESCRIBE_PARAMETER (10002)
			Info Value = 0x7ffd61347bf2
			Buffer Length = 2
			StrLen = 0x7ffd61347bf0
[ODBC][3497][1712343246.059686][SQLGetInfo.c][617]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.059702][SQLGetInfo.c][554]
		Entry:
			Connection = 0x23f12a0
			Info Type = SQL_NEED_LONG_DATA_LEN (111)
			Info Value = 0x7ffd61347bf2
			Buffer Length = 2
			StrLen = 0x7ffd61347bf0
[ODBC][3497][1712343246.059747][SQLGetInfo.c][617]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.059758][SQLAllocHandle.c][540]
		Entry:
			Handle Type = 3
			Input Handle = 0x23f12a0
[ODBC][3497][1712343246.059969][SQLAllocHandle.c][1081]
		Exit:[SQL_SUCCESS]
			Output Handle = 0x24c2a40
[ODBC][3497][1712343246.060007][SQLGetTypeInfo.c][168]
		Entry:
			Statement = 0x24c2a40
			Data Type = SQL_VARCHAR
[ODBC][3497][1712343246.062125][SQLGetTypeInfo.c][318]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.062173][SQLFetch.c][162]
		Entry:
			Statement = 0x24c2a40
[ODBC][3497][1712343246.062197][SQLFetch.c][348]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.062212][SQLGetData.c][237]
		Entry:
			Statement = 0x24c2a40
			Column Number = 3
			Target Type = 4 SQL_INTEGER
			Buffer Length = 4
			Target Value = 0x7ffd61347bf4
			StrLen Or Ind = (nil)
[ODBC][3497][1712343246.062237][SQLGetData.c][501]
		Exit:[SQL_SUCCESS]                
			Buffer = [32767]                
			Strlen Or Ind = NULLPTR
[ODBC][3497][1712343246.062248][SQLFreeStmt.c][144]
		Entry:
			Statement = 0x24c2a40
			Option = 0
[ODBC][3497][1712343246.062391][SQLFreeStmt.c][263]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.062404][SQLFreeHandle.c][381]
		Entry:
			Handle Type = 3
			Input Handle = 0x24c2a40
[ODBC][3497][1712343246.062441][SQLFreeHandle.c][491]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.062452][SQLAllocHandle.c][540]
		Entry:
			Handle Type = 3
			Input Handle = 0x23f12a0
[ODBC][3497][1712343246.062505][SQLAllocHandle.c][1081]
		Exit:[SQL_SUCCESS]
			Output Handle = 0x24c2a40
[ODBC][3497][1712343246.062517][SQLGetTypeInfo.c][168]
		Entry:
			Statement = 0x24c2a40
			Data Type = Unknown(-9)
[ODBC][3497][1712343246.062687][SQLGetTypeInfo.c][318]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.062700][SQLFetch.c][162]
		Entry:
			Statement = 0x24c2a40
[ODBC][3497][1712343246.062713][SQLFetch.c][348]
		Exit:[SQL_NO_DATA]
[ODBC][3497][1712343246.062722][SQLFreeStmt.c][144]
		Entry:
			Statement = 0x24c2a40
			Option = 0
[ODBC][3497][1712343246.062754][SQLFreeStmt.c][263]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.062765][SQLFreeHandle.c][381]
		Entry:
			Handle Type = 3
			Input Handle = 0x24c2a40
[ODBC][3497][1712343246.062786][SQLFreeHandle.c][491]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.062802][SQLAllocHandle.c][540]
		Entry:
			Handle Type = 3
			Input Handle = 0x23f12a0
[ODBC][3497][1712343246.062866][SQLAllocHandle.c][1081]
		Exit:[SQL_SUCCESS]
			Output Handle = 0x24c2a40
[ODBC][3497][1712343246.062878][SQLGetTypeInfo.c][168]
		Entry:
			Statement = 0x24c2a40
			Data Type = SQL_VARBINARY
[ODBC][3497][1712343246.063021][SQLGetTypeInfo.c][318]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.063034][SQLFetch.c][162]
		Entry:
			Statement = 0x24c2a40
[ODBC][3497][1712343246.063047][SQLFetch.c][348]
		Exit:[SQL_NO_DATA]
[ODBC][3497][1712343246.063055][SQLFreeStmt.c][144]
		Entry:
			Statement = 0x24c2a40
			Option = 0
[ODBC][3497][1712343246.063087][SQLFreeStmt.c][263]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.063097][SQLFreeHandle.c][381]
		Entry:
			Handle Type = 3
			Input Handle = 0x24c2a40
[ODBC][3497][1712343246.063113][SQLFreeHandle.c][491]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.063125][SQLAllocHandle.c][540]
		Entry:
			Handle Type = 3
			Input Handle = 0x23f12a0
[ODBC][3497][1712343246.063181][SQLAllocHandle.c][1081]
		Exit:[SQL_SUCCESS]
			Output Handle = 0x24c2a40
[ODBC][3497][1712343246.063195][SQLGetTypeInfo.c][168]
		Entry:
			Statement = 0x24c2a40
			Data Type = SQL_TYPE_TIMESTAMP
[ODBC][3497][1712343246.064045][SQLGetTypeInfo.c][318]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.064071][SQLFetch.c][162]
		Entry:
			Statement = 0x24c2a40
[ODBC][3497][1712343246.064085][SQLFetch.c][348]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.064096][SQLGetData.c][237]
		Entry:
			Statement = 0x24c2a40
			Column Number = 3
			Target Type = 4 SQL_INTEGER
			Buffer Length = 4
			Target Value = 0x7ffd61347bf4
			StrLen Or Ind = (nil)
[ODBC][3497][1712343246.064112][SQLGetData.c][501]
		Exit:[SQL_SUCCESS]                
			Buffer = [29]                
			Strlen Or Ind = NULLPTR
[ODBC][3497][1712343246.064121][SQLFreeStmt.c][144]
		Entry:
			Statement = 0x24c2a40
			Option = 0
[ODBC][3497][1712343246.064168][SQLFreeStmt.c][263]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.064179][SQLFreeHandle.c][381]
		Entry:
			Handle Type = 3
			Input Handle = 0x24c2a40
[ODBC][3497][1712343246.064198][SQLFreeHandle.c][491]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.065500][SQLDisconnect.c][208]
		Entry:
			Connection = 0x23f12a0
[ODBC][3497][1712343246.067156][SQLDisconnect.c][364]
		Exit:[SQL_SUCCESS]
[ODBC][3497][1712343246.067190][SQLFreeHandle.c][284]
		Entry:
			Handle Type = 2
			Input Handle = 0x23f12a0
[ODBC][3497][1712343246.067208][SQLFreeHandle.c][333]
		Exit:[SQL_SUCCESS]
```

**Comment by gordthompson** [comment]
That behaviour looks just the same as described here

https://github.com/mkleehammer/pyodbc/issues/1329#issuecomment-1967500744

except that SQLGetDiagRecW returns an actual error instead of SQL_NO_DATA.

TL;DR - pyodbc_4 doesn't freak out if SQLDriverConnectW returns an error, but pyodbc_5 does.

**Comment by bkline** [comment]
@Deividhp13 I was able to confirm that the DNS resolution failure no longer happens with the latest version of `pyodbc`.

I tested with an RPM-based container, as Cloudera only offers two options for the 64-bit Linux driver: Debian 7 and RHEL, and Debian 7 is so old that I'm pretty sure its Python wouldn't support `pyodbc` 5.x.

### odbc.ini

```ini
[ODBC Data Sources]
Hadoop LDAP=Cloudera ODBC Driver for Impala
[Hadoop LDAP]
Driver=/opt/cloudera/impalaodbc/lib/64/libclouderaimpalaodbc64.so
Host=localhost
Port=25008
AuthMech=User Name and Password
UID=scott
PWD=tiger
```

### odbcinst.ini

```ini
[Cloudera ODBC Driver for Impala]
Driver=/opt/cloudera/impalaodbc/lib/64/libclouderaimpalaodbc64.so
```

### Test script

```python
import pyodbc
print(pyodbc.drivers())
conn = pyodbc.connect("DSN=Hadoop LDAP")
print("Connected!")
```

### Test output

```
['Cloudera ODBC Driver for Impala']
Traceback (most recent call last):
  File "/test.py", line 3, in <module>
    conn = pyodbc.connect("DSN=Hadoop LDAP")
pyodbc.Error: ('HY000', '[HY000] [Cloudera][DriverSupport] (1110) Unexpected response received from server. Please ensure the server host and port specified for the connection are correct and confirm if SSL should be enabled for the connection. (1110) (SQLDriverConnect)')
```

That's the expected failure message, since I haven't actually stood up a Hadoop server. It confirms that we have gotten past the step of locating the DSN definition, which was the failure you reported. Can you test with 5.3.0? If that works in your environment, you can close this ticket.

### Test environment
Component | Version
--- | ---
OS | Rocky Linux 9.3
Python | 3.9.25
Driver | ClouderaImpalaODBC-2.6.11.1011-1.x86_64
Driver Manager | unixODBC-2.3.9-4.el9.x86_64
pyodbc | 5.3.0
